### PR TITLE
page_cache: ensure forward progress on miss

### PR DESCRIPTION
**Pull request by problame**
Problem
=======

Prior to this PR, when we had a cache miss, we'd get back a write guard, fill it, the drop it and retry the read from cache.

If there's severe contention for the cache, it could happen that the just-filled data gets evicted before our retry, resulting in lost work and no forward progress.

Solution
========

This PR leverages the now-available `tokio::sync::RwLockWriteGuard`'s `downgrade()` functionality to turn the filled slot write guard into a read guard.
We don't drop the guard at any point, so, forward progress is ensured.

Refs
====

Stacked atop https://github.com/neondatabase/neon/pull/5480 

part of https://github.com/neondatabase/neon/issues/4743
specifically part of https://github.com/neondatabase/neon/issues/5479

### CODE REVIEW

File: pageserver/src/page_cache.rs [hunk 1]
```diff
@@ -287,44 +287,61 @@ impl AsRef<[u8; PAGE_SZ]> for PageReadGuard<'_> {
...
+        }
     }
 }
 
 impl AsMut<[u8; PAGE_SZ]> for PageWriteGuard<'_> {
     fn as_mut(&mut self) -> &mut [u8; PAGE_SZ] {
-        self.inner.buf
+        match &mut self.state {
+            PageWriteGuardState::Invalid { inner, _permit } => inner.buf,
+            PageWriteGuardState::Downgraded => todo!(),
```
Comment (arpad-m) [inline pageserver/src/page_cache.rs:325]:
```suggestion
            PageWriteGuardState::Downgraded => unreachable!(),
```

**Review by arpad-m — APPROVED** [review]
Very nice, lets us get rid of those loops :+1:

File: pageserver/src/page_cache.rs [hunk 2]
```diff
@@ -287,44 +287,61 @@ impl AsRef<[u8; PAGE_SZ]> for PageReadGuard<'_> {
...
     /// Mark that the buffer contents are now valid.
-    pub fn mark_valid(&mut self) {
-        assert!(self.inner.key.is_some());
-        assert!(
-            !self.valid,
-            "mark_valid called on a buffer that was already valid"
-        );
-        self.valid = true;
+    #[must_use]
+    pub fn mark_valid(mut self) -> PageReadGuard<'a> {
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/page_cache.rs:333]: TIL

[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/page_cache.rs:333]: Ah no, this suggestion is invalid, `Self` is PageWriteGuard but we're returning `PageReadGuard` here.

**Pull request merged**